Lib/logging/handlers.py

```diff
@@ -1015,8 +1015,22 @@ def emit(self, record):
                     self.socket.close()
                     self._connect_unixsocket(self.address)
                     self.socket.send(msg)
+            # UDP socket type 
             elif self.socktype == socket.SOCK_DGRAM:
                 self.socket.sendto(msg, self.address)
+            # TCP socket type
+            # This block doesn't exist in original handlers.py code for TCP 
+            elif self.socktype == socket.SOCK_STREAM:
+                try:
+
+                    self.socket.sendall(msg)
+                # create new TCP socket stream, if TCP socket broken or disconnected 
+                except OSError:
+
+                    self.socket = socket.socket(socket.AF_INET, self.socktype)
+                    self.socket.connect(self.address)
+                    # retry sending msg after making new TCP AF_INET socket
+                    self.socket.sendall(msg)
```

Comment:
And what happens if _this_ `sendall(...)` fails, as the outer one did?

Reply:
Same as the existing block above:

```
                try:
                    self.socket.send(msg)
                except OSError:
                    self.socket.close()
                    self._connect_unixsocket(self.address)
                    self.socket.send(msg)
```
https://github.com/python/cpython/pull/123238/files#diff-0798fd5bd555ae6d7713d10e5266a8bcd7a9800f1fd38d74f09e7bda68a79ebeR1012-R1017

I think it's reasonable to allow this to fail if the attempt connect and resend fails a second time. It means that there's something more than a failed send due to a socket timeout occurring.

What did you have in mind?